### DISCUSSION TITLE
Unnamed AST nodes that might be worth to change to named

### DISCUSSION THREAD

**Discussion by xixiaofinland** [aheber/tree-sitter-sfapex#35] [General]
Here are the unnamed nodes I need to traverse during formatting. The purpose is to list them for analysis to determine whether it makes sense to convert them into named nodes. This doesn't necessarily imply that a fix is required.

**Comment by xixiaofinland** [comment]
`i++` v.s. `i--` and `++i` v.s. `--i` in `update_expression` can't be distinguished as the named node covers `i` only.

**Reply by aheber** [reply]
Does this need to be a named node or should it be a field, or both?

**Reply by xixiaofinland** [reply]
Based on my basic understanding of parsers, a field associated with update_expression seems to make the most sense. 

From a formatting perspective, all the points discussed in this thread don't cause significant issues for me. The only aesthetic concern is that a small number of scenarios need to rely on unnamed nodes, which is slightly inelegant. 

Currently, I look to the left and right of update_expression for any unnamed nodes and simply print them.

**Reply by aheber** [reply]
I'll see how it looks on a couple of these. I think I'm more likely to promote it to a named node and a field so that is can be obvious where the operator is in relation to the expression/identifier.

I'll report back.

**Reply by aheber** [reply]
Proposed change:

```
++i;
i++;
```

would now create a tree of
```
 expression_statement [3, 8] - [3, 12]
    update_expression [3, 8] - [3, 11]
      operator: update_operator [3, 8] - [3, 10]
      operand: identifier [3, 10] - [3, 11]
  expression_statement [4, 8] - [4, 12]
    update_expression [4, 8] - [4, 11]
      operand: identifier [4, 8] - [4, 9]
      operator: update_operator [4, 9] - [4, 11]
```

This gives a named node for `update_operator` and also applies named fields to both the `operator` and `operand`

**Reply by xixiaofinland** [reply]
looks decent!

**Reply by aheber** [reply]
This is available now.

**Reply by xixiaofinland** [reply]
confirm it works!

**Comment by xixiaofinland** [comment]
the `?` and `.` in `ref?.maybe` and `ref?.maybeNot()` don't have associated named nodes.

**Reply by aheber** [reply]
As things stand today:
* `.` generates an anonymous node of `"."`
* `?.` is two anonymous nodes of `"?"` and `"."`

I can understand why this isn't good and should change. I think the question is how?
1. `(safe_navigation_operator "?") "."`
2. `(safe_navigation_operator "?.")`

I think I lean towards option 2 here because I think that is how the language sees it. I tested and wasn't able to leave space between the two characters.

Should we change the stand-alone `"."` to something, or is the presence or absence of the named node enough?

**Reply by xixiaofinland** [reply]
Let's use option 2. [Prettier/Jorjre](https://apex.dangmai.net/) also requires `?` and `.` to sit together.

For `.`, so far I print it by myself according to the context. So the current AST setup is cool for me.

**Reply by aheber** [reply]
OK, I'm good with that. I'll make that change and get it added to the repo.

**Reply by aheber** [reply]
Proposed change:

`System?.debug(a);`

would produce a tree of
```
expression_statement [2, 8] - [2, 25]
  method_invocation [2, 8] - [2, 24]
    object: identifier [2, 8] - [2, 14]
    safe_navigaion_operator [2, 14] - [2, 16]
    name: identifier [2, 16] - [2, 21]
    arguments: argument_list [2, 21] - [2, 24]
      identifier [2, 22] - [2, 23]
```

----------------------

`System.debug(a);`

would produce a tree of
```
expression_statement [2, 8] - [2, 24]
  method_invocation [2, 8] - [2, 23]
    object: identifier [2, 8] - [2, 14]
    name: identifier [2, 15] - [2, 20]
    arguments: argument_list [2, 20] - [2, 23]
      identifier [2, 21] - [2, 22]
```

**Reply by xixiaofinland** [reply]
excellent! R&R

**Reply by xixiaofinland** [reply]
pls let me know whenever this is implemented, so I can update my code. @aheber 
no hurry.

**Reply by aheber** [reply]
Will do, I have the changes on my machine but haven't pushed up or published yet. Probably this weekend.

**Reply by aheber** [reply]
This is available now.

**Reply by xixiaofinland** [reply]
confirm it works!

**Comment by xixiaofinland** [comment]
`accessor_declaration` doesn't have named children nodes to break down `get` and `;`.

example:
```
public class Me {
  public double MyReadWriteProp { get;  }
}
```

**Reply by aheber** [reply]
I don't want to try and add named nodes for `;` unless we really need to. Would like to understand how it would be helpful though. Is there something else we could do that would help you accomplish your goals? Maybe make the body a named field if it is present and it's absence is enough indication?

For `get` and `set` would it be enough to make it a named field even if it were populated by an anonymous node? Does it need to be a named node?

I've gone back and forth on keywords. I recently changed modifiers so they have named nodes though I may take away the quoted anonymous node on those. I haven't decided, the duplication isn't necessary that is for sure.

When the keyword is obvious, something like "public `class` MyClass {}", I don't make `class` a named node because the fact that it parsed correctly ensured it was present. `get` and `set` aren't the same because if you were just parsing tree nodes it could be either keyword, if I gave you a way to easily grab it, whichever it was, would that be sufficient? Should they be promoted so that they are visible in tree nodes?

I don't have a rigid policy here but I'm starting to make one. Would appreciate input especially now that you have use cases that add pressure one way or the other.

**Reply by aheber** [reply]
What I'm thinking for minimum options:

```apex
        get {
            return 1;
        }
        set;
```

Produces a tree of
```
        accessor_list [1, 11] - [6, 5]
          accessor_declaration [2, 5] - [4, 9]
            body: block [2, 9] - [4, 9]
              return_statement [3, 9] - [3, 18]
                int [3, 16] - [3, 17]
          accessor_declaration [5, 8] - [5, 12]
```

If I show anonymous nodes it fleshes out much better
```
        accessor_list [1, 11] - [6, 5]
          "{" [1, 11] - [1, 12]
          accessor_declaration [2, 5] - [4, 9]
            accessor: "get" [2, 5] - [2, 8]
            body: block [2, 9] - [4, 9]
              "{" [2, 9] - [2, 10]
              return_statement [3, 9] - [3, 18]
                "return" [3, 9] - [3, 15]
                int [3, 16] - [3, 17]
                ";" [3, 17] - [3, 18]
              "}" [4, 8] - [4, 9]
          accessor_declaration [5, 8] - [5, 12]
            accessor: "set" [5, 8] - [5, 11]
            ";" [5, 11] - [5, 12]
          "}" [6, 4] - [6, 5]
```

**Reply by aheber** [reply]
(I added a new checkbox to the Playground that will show anonymous nodes, I think it makes it much easier to explore the tree)

**Reply by aheber** [reply]
Thoughts on the proposed tree?

**Reply by xixiaofinland** [reply]
Wait, my bad for unclear message. I think we talk about two different things now. 

I don't need to tell the value difference between `get` and `set`. 
My issue was, the `accessor_declaration` maps to two tokens `get` and `;`, thus I can't remove any potential empty spaces in-between without processing the un-named children, for example in `get    ;` . 

However, this isn't really a big issue at all by given that it's scoped in `accessor_declaration` node only.

my current code is as below. 
```
        // it travsers un-named children
        node.all_children_vec().iter().for_each(|c| {
            if !c.is_named() {
                result.push_str(c.v(source_code));
            }
        });
```

**Reply by xixiaofinland** [reply]
I raised this here just in case it's not aligned with parser design best practice and you want to enhance it. I'm quite comfortable with the current implementation already.

**Reply by aheber** [reply]
My thoughts were grab the new `accessor` name field and print it directly, check for a `body` named field, if it is there then print it, if it is absent then print a `;`. You don't need to see or touch the `";"` anonymous node in this case. 

This should allow you not to worry about space between, unless I don't understand still.

**Reply by xixiaofinland** [reply]
@aheber  what you proposed makes sense, we can go with this 👍

**Reply by aheber** [reply]
I think these changes made it into the latest batch. Give it a run and let me know if I missed anything.

**Reply by xixiaofinland** [reply]
I confirm this works!

**Reply by aheber** [reply]
Thanks for checking.
